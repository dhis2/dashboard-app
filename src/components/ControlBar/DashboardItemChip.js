import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Chip, colors } from '@dhis2/ui-core';
import { Link } from 'react-router-dom';
import debounce from 'lodash/debounce';

import StarIcon from '../../icons/Star';
import { apiPostDataStatistics } from '../../api/dataStatistics';

const styles = {
    chip: {
        margin: '3px',
    },
    link: {
        color: colors.grey600,
        display: 'inline-block',
        textDecoration: 'none',
        verticalAlign: 'top',
    },
    icon: {
        height: '20px',
        marginTop: '2px',
        width: '20px',
    },
    selected: {
        fill: colors.white,
    },
    unselected: {
        fill: colors.grey700,
    },
};

export const DashboardItemChip = ({
    starred,
    selected,
    label,
    dashboardId,
    classes,
}) => {
    const chipProps = {
        selected,
        className: classes.chip,
    };

    if (starred) {
        const selectedState = selected ? classes.selected : classes.unselected;
        chipProps.icon = (
            <StarIcon className={`${classes.icon} ${selectedState}`} />
        );
    }

    return (
        <Link
            className={classes.link}
            to={`/${dashboardId}`}
            onClick={debounce(
                () => apiPostDataStatistics('DASHBOARD_VIEW', dashboardId),
                500
            )}
        >
            <Chip {...chipProps}>{label}</Chip>
        </Link>
    );
};

DashboardItemChip.propTypes = {
    starred: PropTypes.bool.isRequired,
    selected: PropTypes.bool.isRequired,
    label: PropTypes.string.isRequired,
    dashboardId: PropTypes.string.isRequired,
};

export default withStyles(styles)(DashboardItemChip);
