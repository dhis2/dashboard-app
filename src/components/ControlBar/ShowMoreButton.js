import React from 'react';
import i18n from '@dhis2/d2-i18n';
import { colors } from '@dhis2/ui-core';
import { withStyles } from '@material-ui/core/styles';

export const SHOWMORE_BAR_HEIGHT = 16;

const styles = {
    showMore: {
        color: colors.grey700,
        cursor: 'pointer',
        fontSize: 11,
        paddingTop: 5,
        '&:hover': {
            textDecoration: 'underline',
        },
    },
    disabled: {
        paddingTop: 5,
        color: colors.grey500,
        fontSize: 11,
        cursor: 'not-allowed',
    },
};

export const ShowMoreButton = ({ onClick, isMaxHeight, classes, disabled }) => {
    return (
        <div style={{ textAlign: 'center' }}>
            {disabled ? (
                <div className={classes.disabled}>{i18n.t('Show more')}</div>
            ) : (
                <div className={classes.showMore} onClick={onClick}>
                    {isMaxHeight ? i18n.t('Show less') : i18n.t('Show more')}
                </div>
            )}
        </div>
    );
};

export default withStyles(styles)(ShowMoreButton);
