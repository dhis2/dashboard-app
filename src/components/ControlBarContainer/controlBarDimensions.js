import { END_FLAP_HEIGHT } from '@dhis2/d2-ui-core/control-bar/ControlBar';

export const CONTROL_BAR_ROW_HEIGHT = 46;
export const CONTROL_BAR_OUTER_HEIGHT_DIFF = 14;

export const getInnerHeight = rows => {
    return rows * CONTROL_BAR_ROW_HEIGHT;
};

export const getOuterHeight = (rows, expandable) => {
    const flapHeight = !expandable ? END_FLAP_HEIGHT : 0;

    return getInnerHeight(rows) + CONTROL_BAR_OUTER_HEIGHT_DIFF + flapHeight;
};
