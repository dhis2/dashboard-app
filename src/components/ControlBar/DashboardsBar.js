import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import ControlBar from './ControlBar';
import arraySort from 'd2-utilizr/lib/arraySort';

import Chip from './DashboardItemChip';
import AddCircleIcon from '../../icons/AddCircle';
import Filter from './Filter';
import ShowMoreButton from './ShowMoreButton';
import {
    FIRST_ROW_PADDING_HEIGHT,
    MIN_ROW_COUNT,
    getRowsHeight,
    getControlBarHeight,
    getNumRowsFromHeight,
} from './controlBarDimensions';
import { sGetControlBarUserRows } from '../../reducers/controlBar';
import { sGetAllDashboards } from '../../reducers/dashboards';
import { sGetFilterName } from '../../reducers/dashboardsFilter';
import { sGetSelectedId } from '../../reducers/selected';
import { acSetControlBarUserRows } from '../../actions/controlBar';
import { acSetFilterName } from '../../actions/dashboardsFilter';
import { orObject, orArray } from '../../modules/util';
import { apiPostControlBarRows } from '../../api/controlBar';

import classes from './styles/DashboardsBar.module.css';

export const MAX_ROW_COUNT = 10;

export class DashboardsBar extends Component {
    state = {
        rows: MIN_ROW_COUNT,
    };

    setInitialDashboardState = rows => {
        this.setState({ rows, isMaxHeight: rows === MAX_ROW_COUNT });
    };

    componentDidMount() {
        this.setInitialDashboardState(this.props.userRows);
    }

    componentWillReceiveProps(nextProps) {
        this.setInitialDashboardState(nextProps.userRows);
    }

    onChangeHeight = newHeight => {
        const adjustedHeight = newHeight - 52; // don't rush the transition to a bigger row count
        const newRows = Math.max(
            MIN_ROW_COUNT,
            getNumRowsFromHeight(adjustedHeight)
        );

        if (newRows !== this.state.rows) {
            const newRowCount = Math.min(newRows, MAX_ROW_COUNT);

            this.props.onChangeHeight(newRowCount);
        }
    };

    onEndDrag = () => {
        return apiPostControlBarRows(this.state.rows);
    };

    onToggleMaxHeight = () => {
        const rows =
            this.state.rows === MAX_ROW_COUNT
                ? this.props.userRows
                : MAX_ROW_COUNT;

        this.setState({ rows, isMaxHeight: !this.state.isMaxHeight });
    };

    onSelectDashboard = () => {
        this.props.history.push(`/${this.props.dashboards[0].id}`);
    };

    render() {
        const { dashboards, name, selectedId, onChangeFilterName } = this.props;

        const rowCount = this.state.isMaxHeight
            ? MAX_ROW_COUNT
            : this.state.rows;
        const controlBarHeight = getControlBarHeight(rowCount);
        const contentWrapperStyle = {
            padding: `${FIRST_ROW_PADDING_HEIGHT}px 6px 0 6px`,
            overflowY: this.state.isMaxHeight ? 'auto' : 'hidden',
            height: getRowsHeight(rowCount) + FIRST_ROW_PADDING_HEIGHT,
        };

        return (
            <ControlBar
                height={controlBarHeight}
                onChangeHeight={this.onChangeHeight}
                onEndDrag={this.onEndDrag}
                editMode={false}
            >
                <div style={contentWrapperStyle}>
                    <div className={classes.leftControls}>
                        <Link
                            style={{
                                display: 'inline-block',
                                textDecoration: 'none',
                                marginRight: 10,
                                position: 'relative',
                                top: '2px',
                            }}
                            to={'/new'}
                        >
                            <AddCircleIcon />
                        </Link>
                        <Filter
                            name={name}
                            onChangeName={onChangeFilterName}
                            onKeypressEnter={this.onSelectDashboard}
                        />
                    </div>
                    {orArray(dashboards).map(dashboard => (
                        <Chip
                            key={dashboard.id}
                            label={dashboard.displayName}
                            starred={dashboard.starred}
                            dashboardId={dashboard.id}
                            selected={dashboard.id === selectedId}
                        />
                    ))}
                </div>
                <ShowMoreButton
                    onClick={this.onToggleMaxHeight}
                    isMaxHeight={this.state.isMaxHeight}
                    disabled={this.props.userRows === MAX_ROW_COUNT}
                />
            </ControlBar>
        );
    }
}

const mapStateToProps = state => ({
    dashboards: sGetAllDashboards(state),
    name: sGetFilterName(state),
    userRows: sGetControlBarUserRows(state),
    selectedId: sGetSelectedId(state),
});

const mapDispatchToProps = {
    onChangeHeight: acSetControlBarUserRows,
    onChangeFilterName: acSetFilterName,
};
const mergeProps = (stateProps, dispatchProps, ownProps) => {
    const dashboards = Object.values(orObject(stateProps.dashboards));
    const displayDashboards = arraySort(
        dashboards.filter(d =>
            d.displayName.toLowerCase().includes(stateProps.name.toLowerCase())
        ),
        'ASC',
        'displayName'
    );

    return {
        ...stateProps,
        ...ownProps,
        ...dispatchProps,
        dashboards: [
            ...displayDashboards.filter(d => d.starred),
            ...displayDashboards.filter(d => !d.starred),
        ],
    };
};

export default withRouter(
    connect(
        mapStateToProps,
        mapDispatchToProps,
        mergeProps
    )(DashboardsBar)
);
