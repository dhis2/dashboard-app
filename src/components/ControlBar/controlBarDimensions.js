import { SHOWMORE_BAR_HEIGHT } from './ShowMoreButton';

export const CONTROL_BAR_ROW_HEIGHT = 40;
export const FIRST_ROW_PADDING_HEIGHT = 10;
export const MIN_ROW_COUNT = 1;

export const HEADERBAR_HEIGHT = 48;

const CONTROL_BAR_OUTER_HEIGHT_DIFF =
    FIRST_ROW_PADDING_HEIGHT + SHOWMORE_BAR_HEIGHT - 2; // 2 pixel "adjustment"

export const getRowsHeight = rows => {
    return rows * CONTROL_BAR_ROW_HEIGHT;
};

export const getNumRowsFromHeight = height => {
    return Math.floor(
        (height - CONTROL_BAR_OUTER_HEIGHT_DIFF) / CONTROL_BAR_ROW_HEIGHT
    );
};

export const getTopOffset = rows => {
    return HEADERBAR_HEIGHT + getControlBarHeight(rows, false);
};

export const getControlBarHeight = rows => {
    return getRowsHeight(rows) + CONTROL_BAR_OUTER_HEIGHT_DIFF;
};
