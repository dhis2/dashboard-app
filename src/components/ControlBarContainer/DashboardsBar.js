import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import ControlBar from '@dhis2/d2-ui-core/control-bar/ControlBar';
import arraySort from 'd2-utilizr/lib/arraySort';

import Chip from './DashboardItemChip';
import D2IconButton from '../../widgets/D2IconButton';
import Filter from './Filter';
import ShowMoreButton from './ShowMoreButton';
import {
    CONTROL_BAR_ROW_HEIGHT,
    CONTROL_BAR_OUTER_HEIGHT_DIFF,
    getInnerHeight,
    getOuterHeight,
} from './controlBarDimensions';
import { sGetControlBarUserRows } from '../../reducers/controlBar';
import { sGetAllDashboards } from '../../reducers/dashboards';
import { sGetFilterName } from '../../reducers/dashboardsFilter';
import { sGetSelectedId } from '../../reducers/selected';
import { acSetControlBarUserRows } from '../../actions/controlBar';
import { acSetFilterName } from '../../actions/dashboardsFilter';
import { orObject, orArray } from '../../modules/util';
import { apiPostControlBarRows } from '../../api/controlBar';

import './ControlBarContainer.css';

export const MIN_ROW_COUNT = 1;
export const MAX_ROW_COUNT = 10;

export class DashboardsBar extends Component {
    state = {
        rows: MIN_ROW_COUNT,
    };

    setInitialDashboardState = rows => {
        this.setState({ rows, isMaxHeight: rows === MAX_ROW_COUNT });
    };

    componentDidMount() {
        this.setInitialDashboardState(this.props.userRows);
    }

    componentWillReceiveProps(nextProps) {
        this.setInitialDashboardState(nextProps.userRows);
    }

    onChangeHeight = (newHeight, onEndDrag) => {
        const newRows = Math.max(
            MIN_ROW_COUNT,
            Math.floor(
                (newHeight - CONTROL_BAR_OUTER_HEIGHT_DIFF) /
                    CONTROL_BAR_ROW_HEIGHT
            )
        );

        if (newRows !== this.state.rows) {
            const newRowCount = Math.min(newRows, MAX_ROW_COUNT);

            this.props.onChangeHeight(newRowCount);
        }
    };

    onEndDrag = () => {
        return apiPostControlBarRows(this.state.rows);
    };

    onToggleMaxHeight = () => {
        const rows =
            this.state.rows === MAX_ROW_COUNT
                ? this.props.userRows
                : MAX_ROW_COUNT;

        this.setState({ rows, isMaxHeight: !this.state.isMaxHeight });
    };

    onSelectDashboard = () => {
        this.props.history.push(`/${this.props.dashboards[0].id}`);
    };

    render() {
        const { dashboards, name, selectedId, onChangeFilterName } = this.props;

        const rowCount = this.state.isMaxHeight
            ? MAX_ROW_COUNT
            : this.state.rows;
        const controlBarHeight = getOuterHeight(rowCount, true);
        const contentWrapperStyle = {
            padding: '10px 6px 0 6px',
            overflowY: this.state.isMaxHeight ? 'auto' : 'hidden',
            height: getInnerHeight(rowCount),
        };

        return (
            <ControlBar
                height={controlBarHeight}
                onChangeHeight={this.onChangeHeight}
                onEndDrag={this.onEndDrag}
                editMode={false}
                expandable={true}
            >
                <div style={contentWrapperStyle}>
                    <div className="left-controls">
                        <Link
                            style={{
                                display: 'inline-block',
                                textDecoration: 'none',
                                marginRight: 10,
                            }}
                            to={'/new'}
                        >
                            <D2IconButton />
                        </Link>
                        <Filter
                            name={name}
                            onChangeName={onChangeFilterName}
                            onKeypressEnter={this.onSelectDashboard}
                        />
                    </div>
                    {orArray(dashboards).map(dashboard => (
                        <Chip
                            key={dashboard.id}
                            label={dashboard.displayName}
                            starred={dashboard.starred}
                            dashboardId={dashboard.id}
                            selected={dashboard.id === selectedId}
                        />
                    ))}
                </div>
                {this.props.userRows !== MAX_ROW_COUNT ? (
                    <ShowMoreButton
                        onClick={this.onToggleMaxHeight}
                        isMaxHeight={this.state.isMaxHeight}
                    />
                ) : null}
            </ControlBar>
        );
    }
}

const mapStateToProps = state => ({
    dashboards: sGetAllDashboards(state),
    name: sGetFilterName(state),
    userRows: sGetControlBarUserRows(state),
    selectedId: sGetSelectedId(state),
});

const mapDispatchToProps = {
    onChangeHeight: acSetControlBarUserRows,
    onChangeFilterName: acSetFilterName,
};
const mergeProps = (stateProps, dispatchProps, ownProps) => {
    const dashboards = Object.values(orObject(stateProps.dashboards));
    const displayDashboards = arraySort(
        dashboards.filter(d =>
            d.displayName.toLowerCase().includes(stateProps.name.toLowerCase())
        ),
        'ASC',
        'displayName'
    );

    return {
        ...stateProps,
        ...ownProps,
        ...dispatchProps,
        dashboards: [
            ...displayDashboards.filter(d => d.starred),
            ...displayDashboards.filter(d => !d.starred),
        ],
    };
};

export default withRouter(
    connect(
        mapStateToProps,
        mapDispatchToProps,
        mergeProps
    )(DashboardsBar)
);
