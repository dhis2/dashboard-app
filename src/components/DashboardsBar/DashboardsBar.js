import cx from 'classnames'
import PropTypes from 'prop-types'
import React, {
    useState,
    useRef,
    useEffect,
    useCallback,
    createRef,
} from 'react'
import { connect } from 'react-redux'
import { acSetControlBarUserRows } from '../../actions/controlBar'
import { apiPostControlBarRows } from '../../api/controlBar'
import { useWindowDimensions } from '../../components/WindowDimensionsProvider'
import { sGetControlBarUserRows } from '../../reducers/controlBar'
import Content from './Content'
import DragHandle from './DragHandle'
import { getRowsFromHeight } from './getRowsFromHeight'
import ShowMoreButton from './ShowMoreButton'
import classes from './styles/DashboardsBar.module.css'

export const MIN_ROW_COUNT = 1
export const MAX_ROW_COUNT = 10

const DashboardsBar = ({
    userRows,
    updateUserRows,
    expanded,
    onExpandedChanged,
}) => {
    const [dragging, setDragging] = useState(false)
    const [mouseYPos, setMouseYPos] = useState(0)
    const userRowsChanged = useRef(false)
    const ref = createRef()
    const { height } = useWindowDimensions()

    const rootElement = document.documentElement

    useEffect(() => {
        if (mouseYPos === 0) {
            return
        }

        const newRows = Math.max(
            MIN_ROW_COUNT,
            getRowsFromHeight(mouseYPos - 52) // don't rush the transition to a bigger row count
        )

        if (newRows < MAX_ROW_COUNT) {
            onExpandedChanged(false)
        }

        if (newRows !== userRows) {
            updateUserRows(Math.min(newRows, MAX_ROW_COUNT))
            userRowsChanged.current = true
        }
    }, [mouseYPos])

    useEffect(() => {
        rootElement.style.setProperty('--user-rows-count', userRows)
    }, [userRows])

    useEffect(() => {
        const vh = height * 0.01
        rootElement.style.setProperty('--vh', `${vh}px`)
    }, [height])

    useEffect(() => {
        if (!dragging && userRowsChanged.current) {
            apiPostControlBarRows(userRows)
            userRowsChanged.current = false
        }
    }, [dragging, userRowsChanged.current])

    const scrollToTop = () => {
        if (expanded) {
            ref.current.scroll(0, 0)
        }
    }

    const memoizedToggleExpanded = useCallback(() => {
        if (expanded) {
            memoizedCancelExpanded()
        } else {
            scrollToTop()
            onExpandedChanged(!expanded)
        }
    }, [expanded])

    const memoizedCancelExpanded = useCallback(() => {
        scrollToTop()
        onExpandedChanged(false)
    }, [])

    return (
        <div
            className={cx(
                classes.bar,
                expanded ? classes.expanded : classes.collapsed
            )}
            data-test="dashboards-bar"
        >
            <div
                className={classes.container}
                data-test="dashboardsbar-container"
            >
                <div className={classes.content} ref={ref}>
                    <Content
                        onChipClicked={memoizedCancelExpanded}
                        onSearchClicked={memoizedToggleExpanded}
                        expanded={expanded}
                    />
                </div>
                <ShowMoreButton
                    onClick={memoizedToggleExpanded}
                    dashboardBarIsExpanded={expanded}
                    disabled={!expanded && userRows === MAX_ROW_COUNT}
                />
                <DragHandle
                    setDragging={setDragging}
                    onHeightChanged={setMouseYPos}
                />
            </div>
            <div className={classes.spacer} />
        </div>
    )
}

DashboardsBar.propTypes = {
    expanded: PropTypes.bool,
    updateUserRows: PropTypes.func,
    userRows: PropTypes.number,
    onExpandedChanged: PropTypes.func,
}

DashboardsBar.defaultProps = {
    onExpandedChanged: Function.prototype,
}

const mapStateToProps = state => ({
    userRows: sGetControlBarUserRows(state),
})

const mapDispatchToProps = {
    updateUserRows: acSetControlBarUserRows,
}

export default connect(mapStateToProps, mapDispatchToProps)(DashboardsBar)
