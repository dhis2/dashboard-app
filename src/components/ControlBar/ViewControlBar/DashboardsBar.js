import React, { useState, useRef, useEffect, createRef } from 'react'
import { connect } from 'react-redux'
import cx from 'classnames'
import PropTypes from 'prop-types'

import Content from './Content'
import ShowMoreButton from './ShowMoreButton'
import DragHandle from './DragHandle'
import { getRowsFromHeight } from './controlBarDimensions'
import { sGetControlBarUserRows } from '../../../reducers/controlBar'
import { acSetControlBarUserRows } from '../../../actions/controlBar'
import { apiPostControlBarRows } from '../../../api/controlBar'
import { useWindowDimensions } from '../../WindowDimensionsProvider'

import classes from './styles/DashboardsBar.module.css'

export const MIN_ROW_COUNT = 1
export const MAX_ROW_COUNT = 10

const DashboardsBar = ({
    userRows,
    updateUserRows,
    expanded,
    onExpandedChanged,
}) => {
    const [dragging, setDragging] = useState(false)
    const userRowsChanged = useRef(false)
    const ref = createRef()
    const { height } = useWindowDimensions()

    const rootElement = document.documentElement

    const adjustRows = newHeight => {
        const newRows = Math.max(
            MIN_ROW_COUNT,
            getRowsFromHeight(newHeight - 52) // don't rush the transition to a bigger row count
        )

        if (newRows !== userRows) {
            updateUserRows(Math.min(newRows, MAX_ROW_COUNT))
            userRowsChanged.current = true
        }
    }

    useEffect(() => {
        rootElement.style.setProperty('--user-rows-count', userRows)
    }, [userRows])

    useEffect(() => {
        const vh = height * 0.01
        rootElement.style.setProperty('--vh', `${vh}px`)
    }, [height])

    useEffect(() => {
        if (!dragging && userRowsChanged.current) {
            apiPostControlBarRows(userRows)
            userRowsChanged.current = false
        }
    }, [dragging, userRowsChanged.current])

    const scrollToTop = () => {
        if (expanded) {
            ref.current.scroll(0, 0)
        }
    }

    const toggleExpanded = () => {
        if (expanded) {
            cancelExpanded()
        } else {
            scrollToTop()
            onExpandedChanged(!expanded)
        }
    }

    const cancelExpanded = () => {
        scrollToTop()
        onExpandedChanged(false)
    }

    return (
        <div
            className={expanded ? classes.expanded : classes.collapsed}
            data-test="dashboards-bar"
        >
            <div className={cx(classes.container)}>
                <div className={classes.content} ref={ref}>
                    <Content
                        onChipClicked={cancelExpanded}
                        onSearchClicked={toggleExpanded}
                        expanded={expanded}
                    />
                </div>
                <ShowMoreButton
                    onClick={toggleExpanded}
                    dashboardBarIsExpanded={expanded}
                    disabled={!expanded && userRows === MAX_ROW_COUNT}
                />
                <DragHandle
                    setDragging={setDragging}
                    onHeightChanged={adjustRows}
                />
            </div>
            <div className={cx(classes.spacer)} />
        </div>
    )
}

DashboardsBar.propTypes = {
    expanded: PropTypes.bool,
    updateUserRows: PropTypes.func,
    userRows: PropTypes.number,
    onExpandedChanged: PropTypes.func,
}

DashboardsBar.defaultProps = {
    expanded: false,
    onExpandedChanged: Function.prototype,
}

const mapStateToProps = state => ({
    userRows: sGetControlBarUserRows(state),
})

const mapDispatchToProps = {
    updateUserRows: acSetControlBarUserRows,
}

export default connect(mapStateToProps, mapDispatchToProps)(DashboardsBar)
