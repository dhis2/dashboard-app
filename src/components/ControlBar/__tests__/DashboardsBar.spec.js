import React from 'react';
import { shallow } from 'enzyme';

import { getStubContext } from '../../../setupTests';
import { MIN_ROW_COUNT } from '../controlBarDimensions';
import { DashboardsBar, MAX_ROW_COUNT } from '../DashboardsBar';
import ShowMoreButton from '../ShowMoreButton';
import DashboardItemChip from '../DashboardItemChip';
import * as api from '../../../api/controlBar';

describe('DashboardsBar', () => {
    let props;
    let shallowDashboardsBar;
    const dashboardsBar = () => {
        if (!shallowDashboardsBar) {
            shallowDashboardsBar = shallow(<DashboardsBar {...props} />, {
                context: getStubContext(),
            });
        }
        return shallowDashboardsBar;
    };

    beforeEach(() => {
        props = {
            controlsStyle: undefined,
            dashboards: undefined,
            name: undefined,
            userRows: MIN_ROW_COUNT,
            selectedId: undefined,
            isMaxHeight: false,
            onChangeHeight: undefined,
            onEndDrag: undefined,
            onToggleMaxHeight: undefined,
            onChangeFilterName: undefined,
        };
        shallowDashboardsBar = undefined;
    });

    it('renders a DashboardsBar with no items', () => {
        expect(dashboardsBar()).toMatchSnapshot();
    });

    it('does not render ShowMoreButton when userRows is MAX_ROW_COUNT', () => {
        props.userRows = MAX_ROW_COUNT;
        expect(dashboardsBar()).toMatchSnapshot();
    });

    it('renders ShowMoreButton when userRows is less than MAX_ROW_COUNT', () => {
        props.userRows = MAX_ROW_COUNT - 1;
        expect(dashboardsBar()).toMatchSnapshot();
    });

    describe('when ShowMore button is toggled', () => {
        it('sets the correct value for isMaxHeight property', () => {
            const bar = dashboardsBar();
            const btn = bar.find(ShowMoreButton);
            btn.simulate('click');
            const rerenderedBtn = bar.find(ShowMoreButton);

            expect(btn.props().isMaxHeight).not.toEqual(
                rerenderedBtn.props().isMaxHeight
            );
        });
    });

    it('calls the api to post user rows when drag ends', () => {
        const spy = jest.spyOn(api, 'apiPostControlBarRows');

        const controlBar = dashboardsBar();
        controlBar.simulate('endDrag');

        expect(spy).toHaveBeenCalled();
    });

    it('triggers onChangeHeight when controlbar height is changed', () => {
        props.onChangeHeight = jest.fn();
        props.userRows = MAX_ROW_COUNT - 1;
        const dbr = dashboardsBar();

        const newPixelHeight = 200; // should be equivalent to 3 rows
        dbr.simulate('changeHeight', newPixelHeight);
        expect(props.onChangeHeight).toHaveBeenCalled();
        expect(props.onChangeHeight).toHaveBeenCalledWith(3);
    });

    it('does not trigger onChangeHeight when controlbar height is changed to similar value', () => {
        props.onChangeHeight = jest.fn();
        props.userRows = MIN_ROW_COUNT;

        const dbr = dashboardsBar();

        const newPixelHeight = 50; //should result in 1 row, same as current
        dbr.simulate('changeHeight', newPixelHeight);
        expect(props.onChangeHeight).not.toHaveBeenCalled();
    });

    describe('when dashboards are provided', () => {
        beforeEach(() => {
            props.dashboards = [
                {
                    id: 'rainbow123',
                    displayName: 'Rainbow Dash',
                    starred: false,
                },
                {
                    id: 'fluttershy123',
                    displayName: 'Fluttershy',
                    starred: true,
                },
            ];
        });

        it('renders DashboardItemChips for each item on dashboard', () => {
            expect(dashboardsBar()).toMatchSnapshot();
        });

        describe('when selected ID is provided', () => {
            beforeEach(() => {
                props.selectedId = 'fluttershy123';
            });

            it('sets the selected property to true', () => {
                expect(
                    dashboardsBar()
                        .find(DashboardItemChip)
                        .at(1)
                        .props().selected
                ).toBe(true);
            });
        });
    });
});
