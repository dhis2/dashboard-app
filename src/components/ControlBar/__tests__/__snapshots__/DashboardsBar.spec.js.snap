// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardsBar does not render ShowMoreButton when userRows is MAX_ROW_COUNT 1`] = `
<ControlBar
  editMode={false}
  height={424}
  onChangeHeight={[Function]}
  onEndDrag={[Function]}
>
  <div
    style={
      Object {
        "height": 410,
        "overflowY": "auto",
        "padding": "10px 6px 0 6px",
      }
    }
  >
    <div
      className="leftControls"
    >
      <Link
        style={
          Object {
            "display": "inline-block",
            "marginRight": 10,
            "position": "relative",
            "textDecoration": "none",
            "top": "2px",
          }
        }
        to="/new"
      >
        <AddCircleIcon />
      </Link>
      <WithStyles(Filter)
        onKeypressEnter={[Function]}
      />
    </div>
  </div>
  <ShowMoreButton
    disabled={true}
    isMaxHeight={true}
    onClick={[Function]}
  />
</ControlBar>
`;

exports[`DashboardsBar renders ShowMoreButton when userRows is less than MAX_ROW_COUNT 1`] = `
<ControlBar
  editMode={false}
  height={384}
  onChangeHeight={[Function]}
  onEndDrag={[Function]}
>
  <div
    style={
      Object {
        "height": 370,
        "overflowY": "hidden",
        "padding": "10px 6px 0 6px",
      }
    }
  >
    <div
      className="leftControls"
    >
      <Link
        style={
          Object {
            "display": "inline-block",
            "marginRight": 10,
            "position": "relative",
            "textDecoration": "none",
            "top": "2px",
          }
        }
        to="/new"
      >
        <AddCircleIcon />
      </Link>
      <WithStyles(Filter)
        onKeypressEnter={[Function]}
      />
    </div>
  </div>
  <ShowMoreButton
    disabled={false}
    isMaxHeight={false}
    onClick={[Function]}
  />
</ControlBar>
`;

exports[`DashboardsBar renders a DashboardsBar with no items 1`] = `
<ControlBar
  editMode={false}
  height={64}
  onChangeHeight={[Function]}
  onEndDrag={[Function]}
>
  <div
    style={
      Object {
        "height": 50,
        "overflowY": "hidden",
        "padding": "10px 6px 0 6px",
      }
    }
  >
    <div
      className="leftControls"
    >
      <Link
        style={
          Object {
            "display": "inline-block",
            "marginRight": 10,
            "position": "relative",
            "textDecoration": "none",
            "top": "2px",
          }
        }
        to="/new"
      >
        <AddCircleIcon />
      </Link>
      <WithStyles(Filter)
        onKeypressEnter={[Function]}
      />
    </div>
  </div>
  <ShowMoreButton
    disabled={false}
    isMaxHeight={false}
    onClick={[Function]}
  />
</ControlBar>
`;

exports[`DashboardsBar when dashboards are provided renders DashboardItemChips for each item on dashboard 1`] = `
<ControlBar
  editMode={false}
  height={64}
  onChangeHeight={[Function]}
  onEndDrag={[Function]}
>
  <div
    style={
      Object {
        "height": 50,
        "overflowY": "hidden",
        "padding": "10px 6px 0 6px",
      }
    }
  >
    <div
      className="leftControls"
    >
      <Link
        style={
          Object {
            "display": "inline-block",
            "marginRight": 10,
            "position": "relative",
            "textDecoration": "none",
            "top": "2px",
          }
        }
        to="/new"
      >
        <AddCircleIcon />
      </Link>
      <WithStyles(Filter)
        onKeypressEnter={[Function]}
      />
    </div>
    <DashboardItemChip
      dashboardId="rainbow123"
      key="rainbow123"
      label="Rainbow Dash"
      selected={false}
      starred={false}
    />
    <DashboardItemChip
      dashboardId="fluttershy123"
      key="fluttershy123"
      label="Fluttershy"
      selected={false}
      starred={true}
    />
  </div>
  <ShowMoreButton
    disabled={false}
    isMaxHeight={false}
    onClick={[Function]}
  />
</ControlBar>
`;
